lib/Code/TidyAll/Plugin.pm

```diff
@@ -110,34 +126,59 @@ sub postprocess_source {
     return $_[1];
 }
 
-sub diff {
-    return $_[1];
-}
-
 sub process_source_or_file {
-    my ( $self, $source, $basename ) = @_;
+    my ( $self, $orig_source, $basename, $check_only ) = @_;
 
+    my $new_source = $orig_source;
     if ( $self->can('transform_source') ) {
         foreach my $iter ( 1 .. $self->tidyall->iterations ) {
-            $source = $self->transform_source($source);
+            $new_source = $self->transform_source($new_source);
         }
     }
     if ( $self->can('transform_file') ) {
-        my $tempfile = $self->_write_temp_file( $basename, $source );
+        my $tempfile = $self->_write_temp_file( $basename, $new_source );
         foreach my $iter ( 1 .. $self->tidyall->iterations ) {
             $self->transform_file($tempfile);
         }
-        $source = read_file($tempfile);
+        $new_source = read_file($tempfile);
     }
     if ( $self->can('validate_source') ) {
-        $self->validate_source($source);
+        $self->validate_source($new_source);
     }
     if ( $self->can('validate_file') ) {
-        my $tempfile = $self->_write_temp_file( $basename, $source );
+        my $tempfile = $self->_write_temp_file( $basename, $new_source );
         $self->validate_file($tempfile);
     }
 
-    return $source;
+    my $diff;
+    if ( $check_only && $new_source ne $orig_source ) {
+        $diff = $self->_maybe_diff( $orig_source, $new_source, $basename );
+    }
+
+    return ( $new_source, $diff );
+}
+
+sub _maybe_diff {
+    my $self = shift;
+
+    return unless $self->diff_on_tidy_error;
+
+    my $orig     = shift;
+    my $new      = shift;
+    my $basename = shift;
+
+    my $orig_file = $self->_write_temp_file( $basename . '.orig', $orig );
+    my $new_file  = $self->_write_temp_file( $basename . '.new',  $new );
+
+    my $output;
+    run3(
+        [ $self->_diff_cmd, '-u', $orig_file, $new_file ],
```

Comment:
This feels wrong.  I'd like `_diff_cmd` to be a public attribute, and include a data structure that includes the `-u`.  Who knows if someone for some reason wants non universal diffs?  Wants to use a custom Perl script for generating the diffs rather than the standard diff command?